File: C++ Project/src/Formatter.cpp
#include "Formatter.h"

#include <memory>
#include <regex>

#include "exceptions.h"
#include "BMP_Formatter.h"
#include "PAM_Formatter.h"
#include "XML_Formatter.h"

std::unique_ptr<Formatter> Formatter::match_formatter(const std::string & file_path) {
	if (std::regex_match(file_path, std::regex(".*\\.bmp$")))
		return std::make_unique<BMP_Formatter>(file_path);

	if (std::regex_match(file_path, std::regex(".*\\.pam$")))
		return std::make_unique<PAM_Formatter>(file_path);

	if (std::regex_match(file_path, std::regex(".*\\.xml$")))
		return std::make_unique<XML_Formatter>(file_path);

	throw Format_Not_Supported();
}

File: src/formatters/Formatter.java
package formatters;


import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.regex.Pattern;

import image.Image;
import image.Pixel;
abstract public class Formatter {

	@SuppressWarnings("serial")
	public static class FormatNotSupported extends Exception {
		protected String message = "Format not supported.";

		@Override
		public String getMessage() {
			return this.message;
		}
	}

	protected String filePath;

	protected abstract Pixel getPixel(RandomAccessFile reader) throws IOException;

	protected abstract void putPixel(RandomAccessFile writer, final Pixel pixel) throws IOException;

	public Formatter(String filePath) {
		this.filePath = filePath;
	}

	public abstract void read(Image image) throws IOException, FormatNotSupported;

	public abstract void write(final Image image) throws IOException;

	public static Formatter matchFormatter(String filePath) throws FormatNotSupported {
		if (Pattern.matches(".*\\.bmp$", filePath))
			return new BMP_Formatter(filePath);
		if (Pattern.matches(".*\\.pam$", filePath))
			return new PAM_Formatter(filePath);

		if (Pattern.matches(".*\\.xml$", filePath))
			return new XML_Formatter(filePath);

		throw new FormatNotSupported();

	}
	 

}